aliyun-ecs/src/main/java/org/jclouds/aliyun/ecs/features/SecurityGroupApi.java

```diff
@@ -0,0 +1,141 @@
+import org.jclouds.aliyun.ecs.domain.options.PaginationOptions;
+import org.jclouds.aliyun.ecs.filters.FormSign;
+import org.jclouds.collect.IterableWithMarker;
+import org.jclouds.collect.PagedIterable;
+import org.jclouds.collect.internal.Arg0ToPagedIterable;
+import org.jclouds.http.functions.ParseJson;
+import org.jclouds.json.Json;
+import org.jclouds.rest.annotations.Fallback;
+import org.jclouds.rest.annotations.QueryParams;
+import org.jclouds.rest.annotations.RequestFilters;
+import org.jclouds.rest.annotations.ResponseParser;
+import org.jclouds.rest.annotations.SelectJson;
+import org.jclouds.rest.annotations.Transform;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.GET;
+import javax.ws.rs.POST;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.MediaType;
+import java.util.List;
+
+@Consumes(MediaType.APPLICATION_JSON)
+@RequestFilters(FormSign.class)
+@QueryParams(keys = { "Version", "Format", "SignatureVersion", "ServiceCode", "SignatureMethod" },
+             values = {"{" + Constants.PROPERTY_API_VERSION + "}", "JSON", "1.0", "ecs", "HMAC-SHA1"})
+public interface SecurityGroupApi {
+
+   @Named("securityGroup:list")
+   @GET
+   @QueryParams(keys = "Action", values = "DescribeSecurityGroups")
+   @ResponseParser(ParseSecurityGroups.class)
+   @Fallback(Fallbacks.EmptyIterableWithMarkerOnNotFoundOr404.class)
+   IterableWithMarker<SecurityGroup> list(@QueryParam("RegionId") String region, ListSecurityGroupsOptions options);
+
+   @Named("securityGroup:list")
+   @GET
+   @QueryParams(keys = "Action", values = "DescribeSecurityGroups")
+   @ResponseParser(ParseSecurityGroups.class)
+   @Transform(ParseSecurityGroups.ToPagedIterable.class)
+   @Fallback(Fallbacks.EmptyPagedIterableOnNotFoundOr404.class)
+   PagedIterable<SecurityGroup> list(@QueryParam("RegionId") String region);
+
+   @Singleton
+   final class ParseSecurityGroups extends ParseJson<SecurityGroups> {
+
+      @Inject
+      ParseSecurityGroups(final Json json) {
+         super(json, TypeLiteral.get(SecurityGroups.class));
+      }
+
+      static class ToPagedIterable extends Arg0ToPagedIterable<SecurityGroup, ParseSecurityGroups.ToPagedIterable> {
+
+         private final ECSComputeServiceApi api;
+
+         @Inject
+         ToPagedIterable(ECSComputeServiceApi api) {
+            this.api = api;
+         }
+
+         @Override
+         protected Function<Object, IterableWithMarker<SecurityGroup>> markerToNextForArg0(final Optional<Object> arg0) {
+            return new Function<Object, IterableWithMarker<SecurityGroup>>() {
+               @Override
+               public IterableWithMarker<SecurityGroup> apply(Object input) {
+                  String regionId = arg0.get().toString();
+                  ListSecurityGroupsOptions listSecurityGroupsOptions = ListSecurityGroupsOptions.Builder.paginationOptions(PaginationOptions.class.cast(input));
```

Comment:
You are creating a brand new options object and forgetting all other options in the original object. You need to keep it.
Get a builder out from the input object (so it keeps all info), and just override the pagination options, instead of creating a brand new options object. Apply this to all API classes.

Reply:
ok refactored